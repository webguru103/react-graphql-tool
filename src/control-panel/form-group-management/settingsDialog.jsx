import * as React from 'react';
import * as yup from 'yup';
import { Formik } from 'formik';
import classNames from 'classnames';
import { withStyles } from '@material-ui/core/styles';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import TextField from '@material-ui/core/TextField';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import FormLabel from '@material-ui/core/FormLabel';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import Button from '@material-ui/core/Button';
import { withUpdateFormGroup } from './api/formGroup.service';
import { compose } from '../../utility';
import { withDialog } from '../../shared/dialog/withDialog';
import validator from '../../validation/helpers';
import c, { FORM_GROUP_VISIBILITY } from './constants';
import styles from './settingsDialog.styles';

type PropType = {
  classes: Object,
  closeDialog: Function,
  updateFormGroup: Function,
  formGroupId: number,
  formGroupVis: string,
  formGroupName: string,
  onSubmit: Function,
};

type StateType = {
  networkError: React.Element,
}

class SettingsDialogC extends React.PureComponent<PropType, StateType> {

  state = {
    networkError: '',
  };

  submitUpdateFormGroup = async ({
    groupName,
    visibility,
  }: {
      groupName: string,
      visibility: string,
    } = {}) => {
    const {
      updateFormGroup, closeDialog, formGroupId, onSubmit,
    } = this.props;

    const formGroup = {
      formGroupName: groupName,
      visibility,
    };
    try {
      await updateFormGroup({ id: formGroupId, formGroup });
      onSubmit();
      closeDialog();
    } catch (err) {
      this.setState({ networkError: c.REQUEST_ERROR });
    }
  };

  render() {
    const {
      classes, closeDialog, formGroupVis, formGroupName,
    } = this.props;

    const {
      networkError,
    } = this.state;

    // TODO: make sure visibility is within a set array of values
    const validate = (values) => {
      const validationResult = validator(values, {
        groupName: yup.string().trim().required(c.EMPTY_GROUP_NAME),
        visibility: yup.string().matches(/(EVERYONE|OREA|NO_ONE)/, c.EMPTY_VISIBILITY).required(c.EMPTY_VISIBILITY),
      });
      return validationResult.errors;
    };

    return (
      <Formik
        onSubmit={this.submitUpdateFormGroup}
        validate={validate}
        initialValues={{
          groupName: (formGroupName || ''),
          visibility: (formGroupVis || ''),
        }}
        render={({
          values, touched, errors, handleSubmit, handleChange,
        }) => (
          <form onSubmit={handleSubmit}>
            <DialogTitle classes={{ root: classes.dialogTitle }} disableTypography>{c.SETTINGS}</DialogTitle>
            <DialogContent classes={{ root: classes.dialogContent }}>
              <div className={classes.flexInputs}>
                <TextField
                  error={touched && touched.groupName && errors && Boolean(errors.groupName)}
                  fullWidth
                  label="Group Name"
                  id="input"
                  InputProps={{
                    classes: {
                      root: classes.textFieldInputWrapper,
                      input: classes.textFieldInput,
                      error: classes.textFieldInputError,
                    },
                    disableUnderline: true,
                  }}
                  InputLabelProps={{
                    classes: { root: `${classNames(classes.textFieldLabel, classes.label)}` },
                    required: true,
                  }}
                  onChange={handleChange}
                  value={values && values.groupName}
                  name="groupName"
                />
                <FormControl
                  error={errors && Boolean(networkError)}
                >
                  <FormLabel
                    className={classNames(classes.label, classes.textFieldLabel)}
                    focused={false}
                  >
                    Visibility
                  </FormLabel>
                  <RadioGroup
                    aria-label="visibility"
                    name="visibility"
                    classes={{ root: classes.radioGroup }}
                    value={values && values.visibility}
                    onChange={handleChange}
                  >
                    <FormControlLabel
                      data-testid="settings-form-group-visible-everyone"
                      value={FORM_GROUP_VISIBILITY.EVERYONE}
                      control={<Radio classes={{ root: classes.radio }} disableRipple />}
                      label="Everyone"
                    />
                    <FormControlLabel
                      data-testid="settings-form-group-visible-orea"
                      value={FORM_GROUP_VISIBILITY.OREA}
                      control={<Radio classes={{ root: classes.radio }} disableRipple />}
                      label="Limit To OREA"
                    />
                    <FormControlLabel
                      data-testid="settings-form-group-visible-no-one"
                      value={FORM_GROUP_VISIBILITY.NO_ONE}
                      control={<Radio classes={{ root: classes.radio }} disableRipple />}
                      label="No One"
                    />
                  </RadioGroup>
                  <FormHelperText
                    error
                  >
                    {networkError || errors.visibility}
                  </FormHelperText>
                </FormControl>
              </div>
            </DialogContent>
            <DialogActions classes={{ root: classes.dialogActions, action: classes.dialogAction }}>
              <Button variant="raised" color="default" onClick={closeDialog}>{c.CANCEL}</Button>
              <Button
                variant="raised"
                color="primary"
                data-testid="submit-button"
                type="submit"
                disabled={
                  Boolean(networkError
                    || (touched.groupName && errors.groupName)
                    || (touched.visibility && errors.visibility))
                }
              >
                {c.CONFIRM}
              </Button>
            </DialogActions>
          </form>
        )}
      />
    );
  }

}

export default compose(
  withStyles(styles, { withTheme: true }),
  withDialog,
  withUpdateFormGroup,
)(SettingsDialogC);
