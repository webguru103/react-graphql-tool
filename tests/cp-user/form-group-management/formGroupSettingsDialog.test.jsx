/* eslint import/no-extraneous-dependencies: 0 */

import * as React from 'react';
import 'jest-dom/extend-expect';
import { render, fireEvent, cleanup } from 'react-testing-library';
import { withContext } from '../../test-helpers/contextWrapper';
import settingsDialog from '../../../src/control-panel/form-group-management/settingsDialog';

describe('Form Group Settings Dialog', () => {
  afterEach(cleanup);
  const Component = withContext(settingsDialog);
  test('renders', () => {
    const { container } = render(<Component classes={{}} />);
    expect(container).toMatchSnapshot();
  });

  test('can not submit if no fields have been filled', (done) => {
    expect.assertions = 3;
    const handleSubmit = jest.fn();
    const { getByTestId } = render(<Component
      classes={{}}
    />);
    const submitButton = getByTestId('submit-button');

    expect(submitButton).not.toBeDisabled();

    fireEvent(
      submitButton,
      new MouseEvent('click', { bubbles: true, cancelable: true }),
    );

    setTimeout(() => {
      expect(handleSubmit).toHaveBeenCalledTimes(0);
      expect(submitButton).toBeDisabled();
      done();
    }, 0);
  });

  test('can not submit if one field is missing', (done) => {
    expect.assertions = 3;
    const handleSubmit = jest.fn();
    const { getByLabelText, getByTestId } = render(<Component
      classes={{}}
    />);
    const groupNameInput = getByLabelText('Group Name *');
    const submitButton = getByTestId('submit-button');

    fireEvent.change(groupNameInput, {
      target: { value: 'OREA Group' },
    });

    expect(submitButton).not.toBeDisabled();

    fireEvent(
      submitButton,
      new MouseEvent('click', { bubbles: true, cancelable: true }),
    );
    setTimeout(() => {
      expect(handleSubmit).toHaveBeenCalledTimes(0);
      expect(submitButton).toBeDisabled();
      done();
    }, 0);
  });

  test.skip('can submit with all fields', async () => {
    expect.assertions = 3;
    const handleSubmit = jest.fn();
    const { getByLabelText, getByTestId } = render(<Component
      classes={{}}
      onSubmit={handleSubmit}
      formGroupId={1}
      formGroupVis="EVERYONE"
      formGroupName="Test Group"
    />);
    const groupNameInput = getByLabelText('Group Name *');
    const submitButton = getByTestId('submit-button');

    fireEvent.change(groupNameInput, {
      target: { value: 'OREA Group' },
    });

    expect(submitButton).not.toBeDisabled();

    await fireEvent.click(getByLabelText('Limit To OREA'));
    await fireEvent.click(submitButton);

    expect(submitButton).not.toBeDisabled();
  });
});
