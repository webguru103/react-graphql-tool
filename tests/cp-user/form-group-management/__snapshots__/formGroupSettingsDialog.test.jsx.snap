// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Group Settings Dialog renders 1`] = `
<div>
  <form>
    <div
      class="MuiDialogTitle-root-15 DialogCreatorInjector-dialogTitle-2"
    >
      <span>
        Settings
      </span>
    </div>
    <div
      class="MuiDialogContent-root-16 DialogCreatorInjector-dialogContent-3"
    >
      <div>
        <div
          class="MuiFormControl-root-17 MuiFormControl-fullWidth-20"
        >
          <label
            class="MuiFormLabel-root-28 MuiFormLabel-required-33 MuiInputLabel-root-21 DialogCreatorInjector-textFieldLabel-9 DialogCreatorInjector-label-10 MuiInputLabel-formControl-22 MuiInputLabel-animated-25"
            data-shrink="false"
            for="input"
          >
            Group Name
            <span
              class="MuiFormLabel-asterisk-34"
            >
              â€‰*
            </span>
          </label>
          <div
            class="MuiInputBase-root-48 MuiInput-root-35 DialogCreatorInjector-textFieldInputWrapper-6 MuiInputBase-fullWidth-57 MuiInput-fullWidth-42 MuiInputBase-formControl-49 MuiInput-formControl-36"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input-58 MuiInput-input-43 DialogCreatorInjector-textFieldInput-7"
              id="input"
              name="groupName"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="MuiFormControl-root-17"
        >
          <label
            class="MuiFormLabel-root-28 DialogCreatorInjector-label-10 DialogCreatorInjector-textFieldLabel-9"
          >
            Visibility
          </label>
          <div
            aria-label="visibility"
            class="MuiFormGroup-root-65 DialogCreatorInjector-radioGroup-12"
            role="radiogroup"
          >
            <label
              class="MuiFormControlLabel-root-67"
              data-testid="settings-form-group-visible-everyone"
            >
              <span
                class="MuiButtonBase-root-86 MuiIconButton-root-80 MuiSwitchBase-root-76 MuiRadio-root-71 DialogCreatorInjector-radio-13 MuiRadio-colorSecondary-75"
              >
                <span
                  class="MuiIconButton-label-85"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root-89"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <input
                    class="MuiSwitchBase-input-79"
                    name="visibility"
                    type="radio"
                    value="EVERYONE"
                  />
                </span>
              </span>
              <span
                class="MuiTypography-root-98 MuiTypography-body1-107 MuiFormControlLabel-label-70"
              >
                Everyone
              </span>
            </label>
            <label
              class="MuiFormControlLabel-root-67"
              data-testid="settings-form-group-visible-orea"
            >
              <span
                class="MuiButtonBase-root-86 MuiIconButton-root-80 MuiSwitchBase-root-76 MuiRadio-root-71 DialogCreatorInjector-radio-13 MuiRadio-colorSecondary-75"
              >
                <span
                  class="MuiIconButton-label-85"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root-89"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <input
                    class="MuiSwitchBase-input-79"
                    name="visibility"
                    type="radio"
                    value="OREA"
                  />
                </span>
              </span>
              <span
                class="MuiTypography-root-98 MuiTypography-body1-107 MuiFormControlLabel-label-70"
              >
                Limit To OREA
              </span>
            </label>
            <label
              class="MuiFormControlLabel-root-67"
              data-testid="settings-form-group-visible-no-one"
            >
              <span
                class="MuiButtonBase-root-86 MuiIconButton-root-80 MuiSwitchBase-root-76 MuiRadio-root-71 DialogCreatorInjector-radio-13 MuiRadio-colorSecondary-75"
              >
                <span
                  class="MuiIconButton-label-85"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root-89"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <input
                    class="MuiSwitchBase-input-79"
                    name="visibility"
                    type="radio"
                    value="NO_ONE"
                  />
                </span>
              </span>
              <span
                class="MuiTypography-root-98 MuiTypography-body1-107 MuiFormControlLabel-label-70"
              >
                No One
              </span>
            </label>
          </div>
          <p
            class="MuiFormHelperText-root-124 MuiFormHelperText-error-125"
          />
        </div>
      </div>
    </div>
    <div
      class="MuiDialogActions-root-132 DialogCreatorInjector-dialogActions-4"
    >
      <button
        class="MuiButtonBase-root-86 MuiButton-root-134 MuiButton-contained-145 MuiButton-raised-148 MuiDialogActions-action-133 DialogCreatorInjector-dialogAction-5"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label-135"
        >
          <span>
            Cancel
          </span>
        </span>
        <span
          class="MuiTouchRipple-root-160"
        />
      </button>
      <button
        class="MuiButtonBase-root-86 MuiButton-root-134 MuiButton-contained-145 MuiButton-containedPrimary-146 MuiButton-raised-148 MuiButton-raisedPrimary-149 MuiDialogActions-action-133 DialogCreatorInjector-dialogAction-5"
        data-testid="submit-button"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label-135"
        >
          <span>
            Confirm
          </span>
        </span>
        <span
          class="MuiTouchRipple-root-160"
        />
      </button>
    </div>
  </form>
</div>
`;
